pkg/erofs/erofs.go

```diff
@@ -0,0 +1,691 @@
+
+	expected := sb.Checksum
+	sb.Checksum = 0
+
+	buf := &bytes.Buffer{}
+	if err := binary.Write(buf, binary.LittleEndian, sb); err != nil {
+		return fmt.Errorf("failed to marshal super block: %v", err)
+	}
+
+	table := crc32.MakeTable(crc32.Castagnoli)
+	checksum := crc32.Checksum(buf.Bytes(), table)
+
+	const offset = SUPER_BLOCK_OFFSET + SUPER_BLOCK_SIZE
+	tailingBytes := make([]byte, sb.BlockSize()-offset)
+
+	if _, err := i.src.Seek(offset, os.SEEK_SET); err != nil {
+		return fmt.Errorf("failed to reposition offset for super block tailing data: %v", err)
+	}
+
+	if err := binary.Read(i.src, binary.LittleEndian, &tailingBytes); err != nil {
+		return fmt.Errorf("failed to read super block tailing data: %v", err)
+	}
+
+	checksum = ^crc32.Update(checksum, table, tailingBytes)
+	if checksum != expected {
+		return fmt.Errorf("invalid checksum: 0x%x, expected: 0x%x", checksum, expected)
+	}
+
+	return nil
+}
+
+// newInode creates a new inode object for the inode identified by nid.
+func (i *Image) newInode(fileType uint8, nid uint64) (*Inode, error) {
+	inode := &Inode{
+		image:    i,
+		fileType: fileType,
+		nid:      nid,
+	}
+
+	offset := i.sb.NidToOffset(nid)
+
+	var format primitive.Uint16
+	format.UnmarshalBytes(i.bytes[offset:])
+	inode.format = uint16(format)
+
+	var rawBlockAddr uint32
+
+	switch inode.Layout() {
+	case INODE_LAYOUT_COMPACT:
+		inode.inodeSize = INODE_COMPACT_SIZE
+
+		var ino InodeCompact
+		ino.UnmarshalUnsafe(i.bytes[offset:])
+
+		if ino.XattrCount != 0 {
+			return nil, fmt.Errorf("xattr not supported")
+		}
+
+		rawBlockAddr = ino.RawBlockAddr
+
+		inode.size = uint64(ino.Size)
+		inode.nlink = uint32(ino.Nlink)
+		inode.mode = ino.Mode
+		inode.uid = uint32(ino.UID)
+		inode.gid = uint32(ino.GID)
+		inode.mtime = i.sb.BuildTime
+		inode.mtimeNsec = i.sb.BuildTimeNsec
+
+	case INODE_LAYOUT_EXTENDED:
+		inode.inodeSize = INODE_EXTENDED_SIZE
+
+		var ino InodeExtended
+		ino.UnmarshalUnsafe(i.bytes[offset:])
+
+		if ino.XattrCount != 0 {
+			return nil, fmt.Errorf("xattr not supported")
+		}
+
+		rawBlockAddr = ino.RawBlockAddr
+
+		inode.size = ino.Size
+		inode.nlink = ino.Nlink
+		inode.mode = ino.Mode
+		inode.uid = ino.UID
+		inode.gid = ino.GID
+		inode.mtime = ino.Mtime
+		inode.mtimeNsec = ino.MtimeNsec
+
+	default:
+		return nil, fmt.Errorf("unknown inode layout %x", inode.Layout())
+	}
+
+	switch inode.DataLayout() {
+	case INODE_DATALAYOUT_FLAT_INLINE:
+		inode.idataOffset = offset + uint64(inode.inodeSize)
+		fallthrough
+
+	case INODE_DATALAYOUT_FLAT_PLAIN:
+		inode.dataOffset = i.sb.BlockAddrToOffset(rawBlockAddr)
+
+	default:
+		return nil, fmt.Errorf("unknown inode data layout %x", inode.DataLayout())
+	}
+
+	blockSize := uint64(i.sb.BlockSize())
+	inode.blocks = uint32((inode.size + (blockSize - 1)) / blockSize)
+
+	return inode, nil
+}
+
+// Inode represents in-memory inode object.
+//
+// +stateify savable
+type Inode struct {
+	image       *Image
+	dataOffset  uint64
+	idataOffset uint64
```

Comment:
OK, I will change `dataOffset`/`idataOffset` to `dataOff`/`idataOff` and also add a `bool` field to replace the `hasInlineData()` helper function.

But we can’t have just one `dataOff uint64` field since both of `dataOffset` and `idataOffset` of an inode can be valid at the same time, e.g. for large directories. More details can be found in this comment: https://github.com/google/gvisor/pull/9308#discussion_r1316811396

Reply:
Sure. Will drop `hasInlineData` in the next version.